Reported users

HusseinTaha opened the pull request:
Added new endpoint for getting the list of reported-users.
Can you please provide a review and let me know ..

Pending:

- [ ] Tests for the new endpoint
- [ ] Documentation for the new endpoint (readme, and\or api)

Can you please tell me for example what are the tests that I should write ? (scenarios, test cases .. )

Closes #55

j3k0 commented:
Hi Hussein, thanks for the job again!

---

First remark, a bit of contributor's best practice.

Unrelated code should not be re-indented in a PR, it makes the diff harder to read. I can hide the white space changes, but anyway, we try to keep commits minimal.

Also when using `git blame` to understand the behavior or reason for a particular line of code, re-indents will get in the way. So when contributing, be careful to not commit this kind of re-indented lines. It's possible to do some "code cleanup", but we only do that in specific pull requests, when needed and it's rarely a good idea (for the `git blame` reason mentioned).

For example: I'd like to know why `users-api` use the `MAILER_SEND_HTML` environment variable, so I hover the line of code.

<img width="487" alt="Screen Shot 2021-12-06 at 13 26 42" src="https://user-images.githubusercontent.com/191881/144838283-c8823bf6-2204-4592-9b1e-99f1715271fb.png">

With `git blame` (or VS Code GitLens plugin), I know it was added for the `GET /reported-users` endpoint. See the problem? Of course, in this case I converted all code to typescript a year ago, so the interesting part of the git history was lost and I would only see `Convert to typescript` as the last commit instead for this line, at least it wouldn't be misleading.

Is it possible to rewrite the commit? (`git reset` is your friend)

---

Second remark, the endpoint should not include `/auth/:authToken`, it's an admin endpoint, not a user endpoint. Only admin can access it, authentication should be done using the `API_SECRET` key with a query parameter like `?secret=xxx`.

Reading through the code I didn't see it being checked though. Maybe I missed it, anyway, make sure endpoint requires the `secret` (there should be a unit test for that).

---

There should be tests for `reported-users-processor.ts`, `events-latest.ts`.

---

There shouldn't be 1 request per user to check the bans status (`bans.get()`). Assuming we're retrieving the last 10000 events, we'll probably end up doing 10000 parallel requests to the `usermeta` module, like a DDoS attack!

Luckily, the `ganomede-usermeta` module accepts bulk requests (cf https://github.com/j3k0/ganomede-usermeta/#get-useridskeys). Unluckily, the `usermeta.ts` file doesn't allow bulk requests yet, but you could add that as it will be useful for an upcoming task too.

---

```
results = results.slice(0, Math.min(totalItemsTobeReturned, results.length));
```

`Math.min` isn't required (`slice` is safe to use if the `end` parameter is larger than size of array)

---

`latestEventConfig.limit` should be configurable with an environment variable. Something like `NUM_EVENTS_FOR_REPORTED_USERS_REPORT` - verbose!

---

From the name, `events-latest` isn't very clear, as is the `createLatest` function.

Suggestion: `latest-events` and `createLatestEventsClient()` (which would returns a `LatestEventsClient`).

You can be a bit more specific with some module / variable names.

Another example is `blockedUsersApi.latest` which, by quickly reading, seems like is a list of the latest blocked users, when it's actually the function used to fetch the latest events. So `blockedUsersApi.latestEventsClient` might be clearer.

---

I didn't finish the review.

HusseinTaha commented:
Hello, 
I adjust the comments, added the tests for hopefully everything :).
I was careful about what to commit and push..
added get bulk from the user metas to  be used for the banned.
waiting for your review!

HusseinTaha commented:
One more thing, the check for get ban will not be for 10,000 requests ..  the events will  be grouped by users .. 
and will done by reduce to calculate the count of reports .. 
so It's still a parallel requests to check the ban, but it depends on the number of the users actually..

j3k0 commented on tests/test-usermeta.ts:
```diff
@@ -234,6 +234,33 @@ describe("usermeta", function() {
 
+    describe(".getBulk", function() {
+      it("delegates to the jsonClient", function() {
+        usermetaClient.getBulk("username1,username2", ["age"], td.function('callback'));
```
`usermetaClient.getBulk` should take an array of usernames and do the `.join` internally (when needed, to build the URI of the external request). It's strange to pass an array as a comma separated string when we're in a language that supports arrays :). Also, if usernames contain comma, we can still make it work - since we can `encodeURIComponent` each individual username, a username with a comma won't create a weird bug.

j3k0 commented:
Quick review, submitted before I properly tested, so you have it now. I'll be back in an hour if needed.

HusseinTaha commented on tests/test-reported-users-processor.ts:
```diff
@@ -0,0 +1,128 @@
+    it('check for each user if he is banned, and return 0 results', (done) => {
+        td.when(bans!.get(td.matchers.anything(), td.callback)).thenCallback(null, { exists: true });
+        eventsProcessor!("1", [blockedEvent1, blockedEvent2], (error, results: UserReports[] | null) => {
+            expect(error).to.be.equal(null);
+            expect(results?.length).to.be.eql(0);
```
well to be honest, I am not sure which one to use :)
Cause it exists also multi times in the test-users-api.ts, so I am confused now...

j3k0 commented:
I get this error:

```
error TS2307: Cannot find module '../src/blocked-users/get-blocks-api' or its corresponding type declarations.
```

Missing file?

HusseinTaha commented:
but "get-blocks-api" is not part of the reported-users? it exists on another branch ..

j3k0 commented:
> but "get-blocks-api" is not part of the reported-users? it exists on another branch ..

My bad, I just let the example test file in the `tests` directory.
